### PR TITLE
Clean up plugin mess

### PR DESCRIPTION
**Pull request by hypeJunction**

### CODE REVIEW

**Comment by jdalsem** [comment]
LGTM besides all the test failures :)

File: engine/classes/Elgg/Di/ServiceProvider.php [hunk 1]
```diff
@@ -179,7 +186,10 @@ public function __construct(Config $config) {
 			return new \Elgg\Cli($console, $c->hooks);
 		});
 
-		$this->setValue('config', $config);
+		$this->setFactory('config', function (ServiceProvider $sp) use ($config) {
+			$this->initConfig($config, $sp);
```
Comment (jeabakker) [inline engine/classes/Elgg/Di/ServiceProvider.php]: `$this` points to the correct var? or should that be `$sp`?

File: engine/classes/Elgg/Di/ServiceProvider.php [hunk 2]
```diff
@@ -524,4 +525,102 @@ public function __construct(Config $config) {
...
+	/**
+	 * Validate, normalize, fill in missing values, and lock some
+	 *
+	 * @param Config          $config Config
+	 * @param ServiceProvider $sp     Service Provider
+	 *
+	 * @return void
+	 * @throws ConfigurationException
+	 */
+	public function initConfig(Config $config, ServiceProvider $sp) {
```
Comment (jeabakker) [inline engine/classes/Elgg/Di/ServiceProvider.php]: couldn't `$sp` be dropped and simply use `$this`?

File: engine/classes/ElggPlugin.php [hunk 3]
```diff
@@ -869,7 +778,7 @@ public function start($flags) {
 
 			$autoload_file = 'vendor/autoload.php';
 			if ($this->canReadFile($autoload_file)) {
-				Application::requireSetupFileOnce("{$this->path}/{$autoload_file}");
+				Application::requireSetupFileOnce("{$this->getPath()}/{$autoload_file}");
```
Comment (jeabakker) [inline engine/classes/ElggPlugin.php]: I think you get double `/` here as `->getPath()` returns with a trailing `/`

File: engine/classes/ElggPlugin.php [hunk 4]
```diff
@@ -885,7 +794,7 @@ public function start($flags) {
 			$this->activateEntities();
 
 			if ($this->canReadFile('start.php')) {
-				$result = Application::requireSetupFileOnce("{$this->path}/start.php");
+				$result = Application::requireSetupFileOnce("{$this->getPath()}/start.php");
```
Comment (jeabakker) [inline engine/classes/ElggPlugin.php]: I think you get double `/` here as `->getPath()` returns with a trailing `/`

File: engine/classes/ElggPlugin.php [hunk 5]
```diff
@@ -922,20 +831,20 @@ public function start($flags) {
 	 * @return mixed The return value of the included file (or 1 if there is none)
 	 */
 	protected function includeFile($filename) {
-		$filepath = "$this->path/$filename";
+		$filepath = "{$this->getPath()}/$filename";
```
Comment (jeabakker) [inline engine/classes/ElggPlugin.php]: I think you get double `/` here as `->getPath()` returns with a trailing `/`

File: engine/classes/ElggPlugin.php [hunk 6]
```diff
@@ -945,10 +854,12 @@ protected function includeFile($filename) {
 	 * Checks whether a plugin file with the given name exists
 	 *
 	 * @param string $filename The name of the file
+	 *
 	 * @return bool
 	 */
 	protected function canReadFile($filename) {
-		$path = "{$this->path}/$filename";
+		$path = "{$this->getPath()}/$filename";
```
Comment (jeabakker) [inline engine/classes/ElggPlugin.php]: I think you get double `/` here as `->getPath()` returns with a trailing `/`

File: engine/classes/ElggPlugin.php [hunk 7]
```diff
@@ -990,7 +903,7 @@ protected function registerViews() {
 		$views = _elgg_services()->views;
 
 		// Declared views first
-		$file = "{$this->path}/views.php";
+		$file = "{$this->getPath()}/views.php";
```
Comment (jeabakker) [inline engine/classes/ElggPlugin.php]: I think you get double `/` here as `->getPath()` returns with a trailing `/`

File: engine/classes/ElggPlugin.php [hunk 8]
```diff
@@ -1104,21 +1017,19 @@ protected function registerWidgets() {
...
 
 	/**
 	 * Registers the plugin's classes
 	 *
-	 * @throws PluginException
 	 * @return true
 	 */
 	protected function registerClasses() {
-		$classes_path = "$this->path/classes";
+		$classes_path = "{$this->getPath()}/classes";
```
Comment (jeabakker) [inline engine/classes/ElggPlugin.php]: I think you get double `/` here as `->getPath()` returns with a trailing `/`

File: engine/classes/Elgg/Di/ServiceProvider.php (hunk 2 above)
Comment (hypeJunction) [reply engine/classes/Elgg/Di/ServiceProvider.php]: No, needs to be deffered

**Review by jeabakker — CHANGES_REQUESTED** [review]
you make the longest running job longer? What's wrong with the current approach besides it takes too long

**Comment by hypeJunction** [comment]
I am tying to use phpdbg instead of xdebug. The coverage job hasn't completed once yet, because it times out

**Comment by jeabakker** [comment]
> I am tying to use phpdbg instead of xdebug. The coverage job hasn't completed once yet, because it times out

agree on that, but why not leave it in it's own job

**Comment by hypeJunction** [comment]
> agree on that, but why not leave it in it's own job

Because it keeps running in the background for an hour taking up resources from other builds

**Comment by hypeJunction** [comment]
Can someone kill this build https://travis-ci.org/Elgg/Elgg/builds/315309512

**Comment by jeabakker** [comment]
>> agree on that, but why not leave it in it's own job
> 
> Because it keeps running in the background for an hour taking up resources from other builds

hmm the advise from travis is to split more https://docs.travis-ci.com/user/speeding-up-the-build/

**Comment by hypeJunction** [comment]
I don't know how we would combine coverage reports. Maybe we should try using code climate instead of scrutinizer.

**Comment by jeabakker** [comment]
> I don't know how we would combine coverage reports.

that can be done https://scrutinizer-ci.com/docs/tools/external-code-coverage/#merging-coverage-data

**Pull request closed without merging**